### PR TITLE
PR: Add option to insert elements above/below in lists (Variable Explorer)

### LINKED ISSUE #13371: ENH: add "Insert above/below" options for lists in variable explorer
## Problem Description

My team works on Spyder integrating our system as a plugin, and we edit some variables using Variable Explorer(specifically lists). Would be very useful to have the option "insert bellow" when you right click on an item, because the "insert" option, inserts the item in the current position. So if you click in the last item and choose insert, variable explorer inserts it in penultimate position. To have it as the last item, you need to duplicate the last item and edit it(what is not very intuitive).

### What steps reproduce the problem?

1. create any list in the editor, for example: `test_list = [1,2,3,4,5]`
2. Click in Run->Run with "IPython console" and "Variable explorer" active
3. Double-click in test_list variable in "Variable explorer" window.
4. In the opened window, right click on the last item and choose "insert"
5. Type 6 for example and hit enter,
6. Validate that the item is inserted between 4 and 5 items.
7. The final sequence is 1,2,3,4,**6**,5

### What is the expected output? What do you see instead?
I suggest to add the option "insert bellow" when editing lists. How: Changing in collectionseditor.py file, class BaseTableView, setup_menu() method the following:

```
#FROM:
self.insert_action = create_action(self, _("Insert"),
                                           icon=ima.icon('insert'),
                                           triggered=self.insert_item)
#TO:
self.insert_action_above = create_action(self, _("Insert above"),
                                           icon=ima.icon('insert'),
                                           triggered=lambda:self.insert_item(bellow=False))
self.insert_action_bellow = create_action(self, _("Insert bellow"),
                                           icon=ima.icon('insert'),
                                           triggered=lambda:self.insert_item(bellow=True))

#FROM:
 menu_actions = [self.edit_action, self.plot_action, self.hist_action,
                        self.imshow_action, self.save_array_action,
                        self.insert_action, self.remove_action,
                        self.copy_action, self.paste_action,
                        self.view_action,
                        None, self.rename_action, self.duplicate_action,
                        None, resize_action, resize_columns_action]
#TO:
 menu_actions = [self.edit_action, self.plot_action, self.hist_action,
                        self.imshow_action, self.save_array_action,
                        self.insert_action_above, self.insert_action_bellow, self.remove_action,
                        self.copy_action, self.paste_action,
                        self.view_action,
                        None, self.rename_action, self.duplicate_action,
                        None, resize_action, resize_columns_action]

#FROM:
add_actions(self.empty_ws_menu,
                    [self.insert_action, self.paste_action,
                     None, resize_action, resize_columns_action])
#TO:
add_actions(self.empty_ws_menu,
                    [self.insert_action_above, self.insert_action_bellow, self.paste_action,
                     None, resize_action, resize_columns_action])
```

and in insert_item(self) method, change:

```
#FROM:
@Slot()
    def insert_item(self):
#TO:
@Slot()
    def insert_item(self, bellow=True):


#FROM:
        if not index.isValid():
            row = self.source_model.rowCount()
        else:
            row = self.proxy_model.mapToSource(index).row()
#TO:
        if not index.isValid():
            row = self.source_model.rowCount()
        else:
            if bellow:
                row = self.proxy_model.mapToSource(index).row()+1
            else:
                row = self.proxy_model.mapToSource(index).row()
```

### Paste Traceback/Error Below (if applicable)

```python-traceback

PASTE TRACEBACK HERE

```

## Versions

* Spyder version: 4.1.3
* Python version: 3.8.3
* Qt version: 1.9.0
* PyQt version: 5.12.3
* Operating System name/version: 

### Dependencies

```
PASTE DEPENDENCIES HERE
```

### PR DESCRIPTION
**Pull request by dpturibio**
## Description of Changes

* [X] Wrote at least one-line docstrings (for any new functions)
* [X] Added unit test(s) covering the changes (if testable)

* [X] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

Added option **insert above/bellow** in variable explorer(to add new line when editing a list for example) as there was only the option **insert**, and the native behavior was to insert above. So to add new line in the end of a list, you had to duplicate that line and edit it(which is not very intuitive). 
isseuSpyderInsertAboveBellow

### Issue(s) Resolved

Fixes #13371 

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:
dpturibio/Diego Prosperi Turibio

### CODE REVIEW

**Comment by juanis2112** [comment]
Thanks a lot for opening this pull request! We'd like to include your contribution in our next release in the Spyder 4 series (Our next release 4.2.0 will be in September). For that, you would need to make your branch to derive from our `4.x` branch (instead of master). You can use the following commands:

```
git checkout 4.x
git checkout enh_branch
git rebase --onto 4.x master enh_branch
git push -f origin enh_branch
```

Let me know if you run into any problems doing this or if you rather have your work be part of Spyder 5, to be released in 2021.

**Comment by dpturibio** [comment]
Thank you @juanis2112 for your comments. I ran the commands you suggested. Let me know if you need additional changes.
Again, thanks a lot for your support.

File: spyder/plugins/variableexplorer/widgets/collectionseditor.py [hunk 1]
```diff
@@ -522,7 +522,8 @@ def __init__(self, parent):
         self.hist_action = None
         self.imshow_action = None
         self.save_array_action = None
-        self.insert_action = None
+        self.insert_action_above = None
+        self.insert_action_bellow = None
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:526]: bellow -> below

File: spyder/plugins/variableexplorer/widgets/collectionseditor.py [hunk 2]
```diff
@@ -583,9 +584,14 @@ def setup_menu(self, minmax):
...
-                                           icon=ima.icon('insert'),
-                                           triggered=self.insert_item)
+        self.insert_action_above = create_action(
+                      self, _("Insert above"),
+                      icon=ima.icon('insert'),
+                      triggered=lambda: self.insert_item(bellow=False))
+        self.insert_action_bellow = create_action(
+                      self, _("Insert bellow"),
+                      icon=ima.icon('insert'),
+                      triggered=lambda: self.insert_item(bellow=True))
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:594]: bellow -> below

File: spyder/plugins/variableexplorer/widgets/collectionseditor.py [hunk 3]
```diff
@@ -607,18 +613,19 @@ def setup_menu(self, minmax):
         menu = QMenu(self)
         menu_actions = [self.edit_action, self.plot_action, self.hist_action,
                         self.imshow_action, self.save_array_action,
-                        self.insert_action, self.remove_action,
-                        self.copy_action, self.paste_action,
-                        self.view_action,
+                        self.insert_action_above, self.insert_action_bellow,
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:616]: bellow -> below

File: spyder/plugins/variableexplorer/widgets/collectionseditor.py [hunk 4]
```diff
@@ -913,16 +920,22 @@ def rename_item(self, new_name=None):
         self.copy_item(erase_original=True, new_name=new_name)
 
     @Slot()
-    def insert_item(self):
+    def insert_item(self, bellow=True):
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:923]: bellow -> below

File: spyder/plugins/variableexplorer/widgets/collectionseditor.py [hunk 5]
```diff
@@ -913,16 +920,22 @@ def rename_item(self, new_name=None):
...
-    def insert_item(self):
+    def insert_item(self, bellow=True):
         """Insert item"""
         index = self.currentIndex()
         if not index.isValid():
             row = self.source_model.rowCount()
         else:
             if self.proxy_model:
-                row = self.proxy_model.mapToSource(index).row()
+                if bellow:
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:930]: bellow -> below

File: spyder/plugins/variableexplorer/widgets/collectionseditor.py [hunk 6]
```diff
@@ -913,16 +920,22 @@ def rename_item(self, new_name=None):
...
+                    row = self.proxy_model.mapToSource(index).row()+1
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:931]: Please add spaces around the `+` sign here.

File: spyder/plugins/variableexplorer/widgets/collectionseditor.py [hunk 7]
```diff
@@ -913,16 +920,22 @@ def rename_item(self, new_name=None):
...
+                else:
+                    row = self.proxy_model.mapToSource(index).row()
             else:
-                row = index.row()
+                if bellow:
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:935]: bellow -> below

File: spyder/plugins/variableexplorer/widgets/collectionseditor.py [hunk 8]
```diff
@@ -913,16 +920,22 @@ def rename_item(self, new_name=None):
...
+                    row = index.row()+1
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:936]: Please add spaces around the `+` sign here.

File: spyder/plugins/variableexplorer/widgets/collectionseditor.py [hunk 9]
```diff
@@ -1228,7 +1241,8 @@ def refresh_menu(self):
                     and not self.readonly
         self.edit_action.setEnabled( condition )
         self.remove_action.setEnabled( condition )
-        self.insert_action.setEnabled( not self.readonly )
+        self.insert_action_above.setEnabled(not self.readonly)
+        self.insert_action_bellow.setEnabled(not self.readonly)
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:1245]: bellow -> below

**Comment by ccordoba12** [comment]
@dpturibio, thanks a lot for your changes! And don't worry about the failing tests, they are unrelated to your code.

@juanis2112, please give @dpturibio's work a bit of manual testing on Binder. If you find everything is working as expected, then approve it. Thanks!

**Comment by dpturibio** [comment]
Thank you for the support so far @ccordoba12. 
@juanis2112 let me know if you need anything from my side.

**Comment by juanis2112** [comment]
@ccordoba12 and @dpturibio I reviewed this and the functionality is working fine. Very good job @dpturibio. However, I just realized that this was implemented for all elements in the variable explorer (including the pane itself) and for most of them it doesn't really make sense to have the "insert above", "insert below" options because they are ordered by default (which is the case of the dictionaries). I suggest only having this two buttons in the case that the element is a list. Otherwise leave the "insert" button one which inserts it according to some default order.
What do you think @ccordoba12 and @dpturibio ?

**Comment by ccordoba12** [comment]
> I suggest only having this two buttons in the case that the element is a list. Otherwise leave the "insert" button one which inserts it according to some default order.

I totally agree with your suggestion @juanis2112. The changes in this PR only make sense for lists, not for other kind of objects.

**Comment by dpturibio** [comment]
Agree with you @juanis2112, thank you very much for this observation. Made the changes for new features be applied just for lists. It worked well in my tests. Let me know if it works for you.

File: spyder/plugins/variableexplorer/widgets/collectionseditor.py [hunk 10]
```diff
@@ -702,8 +715,12 @@ def refresh_plot_entries(self, index):
         else:
             is_array = condition_plot = condition_imshow = is_list \
                      = condition_hist = False
+        is_list_instance =  isinstance(self.source_model.get_data(), list)
```
Comment (juanis2112) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:718]:
```suggestion
        is_list_instance = isinstance(self.source_model.get_data(), list)

```

Comment (juanis2112) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:718]:
```suggestion
        is_list_instance = isinstance(self.source_model.get_data(), list)

```

File: spyder/plugins/variableexplorer/widgets/collectionseditor.py [hunk 11]
```diff
@@ -1229,6 +1252,8 @@ def refresh_menu(self):
         self.edit_action.setEnabled( condition )
         self.remove_action.setEnabled( condition )
         self.insert_action.setEnabled( not self.readonly )
+        self.insert_action_above.setEnabled( not self.readonly )
```
Comment (juanis2112) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:1255]:
```suggestion
        self.insert_action_above.setEnabled(not self.readonly)

```

File: spyder/plugins/variableexplorer/widgets/collectionseditor.py [hunk 12]
```diff
@@ -1229,6 +1252,8 @@ def refresh_menu(self):
         self.edit_action.setEnabled( condition )
         self.remove_action.setEnabled( condition )
         self.insert_action.setEnabled( not self.readonly )
+        self.insert_action_above.setEnabled( not self.readonly )
+        self.insert_action_below.setEnabled( not self.readonly )
```
Comment (juanis2112) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:1256]:
```suggestion
        self.insert_action_below.setEnabled(not self.readonly)

```

**Comment by juanis2112** [comment]
Thanks for fixing this @dpturibio! I tested this and it works fine for me. 
There's one weird thing though: For some reason `is_list_instance` returns `True` for sets which causes sets to have both options but as sets are not editable they are deactivated so I don't really see a problem on that.

Screen Shot 2020-08-06 at 10 06 00 PM

I ran the tests locally and they work fine, not sure what's going on. @ccordoba12 ideas?
Other than that, I'll say this is ready to merge after fixing the styling issues.

**Comment by dpturibio** [comment]
@juanis2112 do you think we can proceed with the merge? Let me know if you need anything from my side, and I will be glad to help.
Thanks in advance.

File: spyder/plugins/variableexplorer/widgets/collectionseditor.py [hunk 13]
```diff
@@ -583,9 +585,18 @@ def setup_menu(self, minmax):
                                                icon=ima.icon('filesave'),
                                                triggered=self.save_array)
         self.save_array_action.setVisible(False)
-        self.insert_action = create_action(self, _("Insert"),
-                                           icon=ima.icon('insert'),
-                                           triggered=self.insert_item)
+        self.insert_action = create_action(
+                      self, _("Insert"),
+                      icon=ima.icon('insert'),
+                      triggered=lambda: self.insert_item(below=False))
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/collectionseditor.py:591]: Please rewrite this code as follows:

```python
self.insert_action = create_action(
    self, _("Insert"),
    icon=ima.icon('insert'),
    triggered=lambda: self.insert_item(below=False)
)
```

And do the same for the actions below this one.

Comment (dpturibio) [reply spyder/plugins/variableexplorer/widgets/collectionseditor.py]: Thank you @ccordoba12 for your inputs. It seems that these changes triggered more fails in automated tests.
Let me know if you want further changes.

Comment (dpturibio) [reply spyder/plugins/variableexplorer/widgets/collectionseditor.py]: @ccordoba12 I did the changes that you mentioned, but 5 tests failed. Is it something that i need to worry about?

Comment (ccordoba12) [reply spyder/plugins/variableexplorer/widgets/collectionseditor.py]: I'm not sure. For now please merge with the latest `4.x` branch to get the fixes to some of our tests.

**Comment by dpturibio** [comment]
@ccordoba12, now all checks have passed. Let me know if everything is ok. Thank you for the support.

**Review by ccordoba12 — APPROVED** [review]
Looks good to me now, thanks @dpturibio!

**Pull request merged**